src/card.rs

```diff
@@ -0,0 +1,260 @@
+        }
+    };
+}
+
+#[derive(PartialEq, Debug)]
+pub enum CardError {
+    Unlocked = 1,
+    Ready = 0,
+    Busy = -1,
+    Wrongdevice = -2,
+    Nocard = -3,
+    Nofile = -4,
+    IOerror = -5,
+    Broken = -6,
+    Exist = -7,
+    Noent = -8,
+    Insspace = -9,
+    Noperm = -10,
+    Limit = -11,
+    Nametoolong = -12,
+    Encoding = -13,
+    Canceled = -14,
+    FatalError = -128,
+}
+
+impl From<i32> for CardError {
+    fn from(i: i32) -> CardError {
+        match i {
+            1 => CardError::Unlocked,
+            0 => CardError::Ready,
+            -1 => CardError::Busy,
+            -2 => CardError::Wrongdevice,
+            -3 => CardError::Nocard,
+            -4 => CardError::Nofile,
+            -5 => CardError::IOerror,
+            -6 => CardError::Broken,
+            -7 => CardError::Exist,
+            -8 => CardError::Noent,
+            -9 => CardError::Insspace,
+            -10 => CardError::Noperm,
+            -11 => CardError::Limit,
+            -12 => CardError::Nametoolong,
+            -13 => CardError::Encoding,
+            -14 => CardError::Canceled,
+            -128 => CardError::FatalError,
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CardInfo {
+    chan: i32,
+    file_num: i32,
+    offset: i32,
+    length: i32,
+    i_block: u16,
+}
+
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CardStat {
+    // Read-only (Set by CARDGetStatus)
+    file_name: [u8; FILE_NAME_MAX],
+    length: u32,
+    game_name: [u8; 4],
+    company: [u8; 2],
+
+    // Read/Write (Set by CARDGetStatus/CardSetStatus)
+    banner_format: u8,
+    icon_addr: u32,
+    icon_format: u16,
+    icon_speed: u16,
+    commend_addr: u32,
+
+    // Read-only (Set by CARDGetStatus)
+    offset_banner: u32,
+    offset_banner_tlut: u32,
+    offset_icon: [u32; CARD_ICON_MAX],
+    offset_icon_tlut: u32,
+    offset_data: u32,
+}
+
+extern "C" {
+    #[link_name = "CARDOpen"]
+    pub fn card_open(chan: i32, name: *const u8, card_info: *mut CardInfo) -> i32;
+    #[link_name = "CARDRead"]
+    pub fn card_read(card_info: *mut CardInfo, data: *mut u8, len: i32, offset: i32) -> i32;
+    #[link_name = "CARDClose"]
+    pub fn card_close(card_info: *mut CardInfo) -> i32;
+    #[link_name = "CARDCreate"]
+    pub fn card_create(chan: i32, name: *const u8, size: u32, card_info: *mut CardInfo) -> i32;
+    #[link_name = "CARDProbeEx"]
+    pub fn card_probe_ex(chan: i32, card_size: *mut i32, sector_size: *mut i32) -> i32;
+    #[link_name = "CARDGetStatus"]
+    pub fn card_get_status(chan: i32, file_no: i32, card_stat: *mut CardStat) -> i32;
+    #[link_name = "CARDWrite"]
+    pub fn card_write(card_info: *mut CardInfo, data: *const u8, length: i32, offset: i32) -> i32;
+}
+
+type Result<T> = result::Result<T, CardError>;
+
+pub struct Card<'a> {
+    card_info: CardInfo,
+    sector_size: i32,
+    _card_size: i32,
+    _save_size: i32,
```

Comment:
What are these for if they are unused?

Reply:
The card_size was just leftover that I forgot to remove, however the save_size should be the size of the save based on the sector size (per CARD api the ammount of bytes writen to the card needs to be a multiple of sector size), but I couldn't figure out how to dynamically create an ArrayVec, only with a const size. And it turns out that 9/10 memcards are 8k sector so it shouldn't really be an issue.